File: leetcode/python/passedProblems/4-median-of-two-sorted-arrays.py
# There are two sorted arrays nums1 and nums2 of size m and n respectively. Find the median of the two sorted arrays. The overall run time complexity should be O(log (m+n)).

class Solution(object):
    def findMedianSortedArrays(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: float
        """
        l1 = len(nums1)
        l2 = len(nums2)
        ak = self.findK(nums1, nums2, (l1+l2+2)/2);
        ak1 = self.findK(nums1, nums2, (l1+l2+1)/2);
#        print ak , ".", ak1;
        return (0.0 + ak + ak1)/2;
    
    def findK(self, nums1, nums2, k):
#        print nums1, ".", nums2, ".", k;
        l1 = len(nums1);
        l2 = len(nums2);
        if l1 == 0:
            return nums2[k-1];
        elif l2 == 0:
            return nums1[k-1];
        if k == 1:
            return min(nums1[0], nums2[0]);
        p1 = max(0, min(l1- 1, (k-2)/2));
        p2 = k - p1 - 2;
        if(p2 > len(nums2)-1):
            p2 = len(nums2)-1;
            p1 = k - p2 - 2;
        if nums1[p1] > nums2[p2] :
            return self.findK(nums1[0:p1+1], nums2[p2+1:], k - p2 -1);
        elif  nums1[p1] == nums2[p2]:
            return nums1[p1];
        else:
            return self.findK(nums1[p1+1:], nums2[0:p2+1], k - p1 - 1);
        
        
s = Solution();
print s.findMedianSortedArrays([1,2,3,5,7,23,25],[2,4,5,6,7,8,9]);

File: leetcode/cpp/4-median-of-two-sorted-arrays.cpp
/**
 * 在2个排序的数组中找中位数|
 */

class Solution {
    int sz1, sz2;
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        sz1 = nums1.size(), sz2 = nums2.size();
        int am1 = findK((sz1+sz2+1)/2, nums1, nums2);
        int am2 = findK((sz1+sz2)/2+1, nums1, nums2);
        return (0.0+am1+am2)/2;
    }
    int findK(int k, vector<int>& n1, vector<int>& n2){
        int a1 = 0, b1 = sz1-1;
        int a2 = 0, b2 = sz2-1;
        while(k >= 0){
            if(a1 > b1) return n2[a2+k-1];
            if(a2 > b2) return n1[a1+k-1];

            int m1 = (b1+a1)/2;
            if(m1-a1+1>=k){
                if(m1 - a1 + 1 == k && n1[m1] <= n2[a2])
                    return n1[m1];
                b1 = m1-1;continue;
            }
            int m2 = k-(m1-a1+1)+a2-1;
            if(m2 > b2){
                k -= m1-a1+1;
                a1 = m1 +1;
                continue;
            }
            if(n1[m1] > n2[m2]){
                b1 = m1;
                k -= m2-a2+1;
                a2 = m2+1;
            }
            else{
                k -= m1-a1+1;
                a1 = m1+1;
                b2 = m2;
            }

        }
    }
};
